app/services/ai_backend/gemini.rb

```diff
@@ -0,0 +1,116 @@
+class AIBackend::Gemini < AIBackend
+  include Tools
+
+  # Rails system tests don't seem to allow mocking because the server and the
+  # test are in separate processes.
+  #
+  # In regular tests, mock this method or the TestClients::Gemini class to do
+  # what you want instead.
+  def self.client
+    if Rails.env.test?
+      ::TestClient::Gemini
+    else
+      ::Gemini
+    end
+  end
+
+  def initialize(user, assistant, conversation = nil, message = nil)
+    super(user, assistant, conversation, message)
+    begin
+      raise ::OpenAI::ConfigurationError if assistant.api_service.requires_token? && assistant.api_service.effective_token.blank?
+      Rails.logger.info "Connecting to Gemini API server at #{assistant.api_service.url} with access token of length #{assistant.api_service.effective_token.to_s.length}"
+      @client = self.class.client.new(credentials: {service: "generative-language-api",
+                                                    api_key: assistant.api_service.effective_token,
+                                                    version: "v1beta"},
+                                      options: { model: assistant.language_model.api_name,
+                                      server_sent_events: true })
+    rescue ::Faraday::UnauthorizedError => e
+      raise OpenAI::ConfigurationError
+    end
+  end
+
+  def client_method_name
+    :stream_generate_content
+  end
+
+  def configuration_error
+    ::OpenAI::ConfigurationError
+  end
+
+  def set_client_config(config)
+    super(config)
+
+    @client_config = {
+      contents: config[:messages],
+      system_instruction: ( system_message(config[:instructions]) if @assistant.language_model.supports_system_message?)
+    }.compact
+  end
+
+  def get_oneoff_message(instructions, messages, params = {})
+    set_client_config(
+      messages: preceding_conversation_messages,
+      instructions: full_instructions,
+    )
+
+    response = @client.send(client_method_name, @client_config)
+    response.dig("candidates",0,"content","parts",0,"text")
+  end
```

Comment:
Do we need this method inside here? I think you should remove this method and just update `get_oneoff_message` inside `ai_backend`. I believe all you need to do is add another `||` with this new dig:

```
response.dig("candidates",0,"content","parts",0,"text")
```